Sources/OpenAPIRuntime/Multipart/MultipartFramesToRawPartsSequence.swift

```diff
@@ -0,0 +1,380 @@
+    ///
+    /// It enforces that when a new part is emitted by the outer sequence, that the new part's body is then fully
+    /// consumed before the outer sequence is asked for the next part.
+    ///
+    /// This is required as the source of bytes is a single stream, so without the current part's body being consumed,
+    /// we can't move on to the next part.
+    actor SharedIterator {
+
+        /// The upstream source of frames.
+        private var upstream: Upstream.AsyncIterator
+
+        /// The underlying state machine.
+        private var stateMachine: StateMachine
+
+        /// Creates a new iterator.
+        /// - Parameter makeUpstreamIterator: A closure that creates the upstream source of frames.
+        init(makeUpstreamIterator: @Sendable () -> Upstream.AsyncIterator) {
+            let upstream = makeUpstreamIterator()
+            self.upstream = upstream
+            self.stateMachine = .init()
+        }
+
+        /// An error thrown by the shared iterator.
+        struct IteratorError: Swift.Error, CustomStringConvertible, LocalizedError {
+
+            /// The underlying error emitted by the state machine.
+            let error: StateMachine.ActionError
+
+            var description: String {
+                switch error {
+                case .partSequenceNextCalledBeforeBodyWasConsumed:
+                    return
+                        "The outer part sequence was asked for the next element before the current part's inner body sequence was fully consumed."
+                case .receivedBodyChunkInInitial:
+                    return
+                        "Received a body chunk from the upstream sequence as the first element, instead of header fields."
+                case .receivedBodyChunkWhenWaitingForHeaders:
+                    return "Received a body chunk from the upstream sequence when expecting header fields."
+                case .receivedFrameWhenAlreadyHasUnsentHeaders:
+                    return "Received another frame before the current frame with header fields was written out."
+                }
+            }
+
+            var errorDescription: String? { description }
+        }
+
+        /// Request the next element from the outer part sequence.
+        /// - Parameter bodyClosure: The closure invoked to fetch the next byte chunk of the part's body.
+        /// - Returns: The next element, or `nil` if finished.
+        /// - Throws: When a parsing error is encountered.
+        func nextFromPartSequence(bodyClosure: @escaping @Sendable () async throws -> ArraySlice<UInt8>?) async throws
+            -> Element?
+        {
+            switch stateMachine.nextFromPartSequence() {
+            case .returnNil: return nil
+            case .fetchFrame:
+                var upstream = upstream
+                let frame = try await upstream.next()
+                self.upstream = upstream
```

Comment:
This dance seems odd. Because of it being an `actor`?

Reply:
Yes, this is one of the issues when using async sequences from actors.

Weird indeed.